File: 0199.二叉树的右视图/0199-二叉树的右视图.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> result=new ArrayList();
        if (root==null) return result;
        rightOrder(root,0,result,0);
        return result;
    }
    private int rightOrder(TreeNode root,int depth,List result,int max){
        if (root==null) return max;
        depth++;
        if (depth>max){
            max=depth;
            result.add(root.val);
        }
        max=Math.max(rightOrder(root.right,depth,result,max),max);
        max=Math.max(rightOrder(root.left,depth,result,max),max);
        return max;
    }
}

File: 0199.二叉树的右视图/0199-二叉树的右视图.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def rightSideView(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        def dfs(root,max_depth,cur,path):
            cur+=1
            if root==None:
                return max_depth
            if (cur>max_depth):
                max_depth=cur
                path.append(root.val)
            max_depth=max(max_depth,dfs(root.right,max_depth,cur,path))
            max_depth=max(max_depth,dfs(root.left,max_depth,cur,path))
            return max_depth
        if root==None:return []
        path=[]
        dfs(root,0,0,path)
        return path
